scripts/named_entity_recognition/code/algorithms/cnn_bilstm_crf.py

```diff
@@ -0,0 +1,113 @@
+# coding: utf-8
+
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# A Character Convolution + BiLSTM + CRF network for NER
+
+# @author：kenjewu
+# @date：2018/12/12
+
+
+import mxnet as mx
+from mxnet import nd
+from mxnet.gluon import nn, rnn
+from .crf import CRF
+
+
+class CNN_BILSTM_CRF(nn.Block):
+    def __init__(self, nchars, nchar_embed, nwords, nword_embed, nfilters, kernel_size,
+                 nhiddens, nlayers, ntag_space, emb_drop_prob, rnn_drop_prob, out_drop_prob,
+                 tag2idx, **kwargs):
+        super(CNN_BILSTM_CRF, self).__init__(**kwargs)
+        with self.name_scope():
+            self.char_embedding_layer = nn.Embedding(nchars, nchar_embed)
+            self.word_embedding_layer = nn.Embedding(nwords, nword_embed)
+
+            # tanh activation or other?
```

Comment:
Why don't we reuse `convolutional_encoder` instead of writing our own character embedding? https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/model/convolutional_encoder.py

Reply:
Yes, this is a better idea to make the code cleaner. I will update it.

Reply:
I have switched to conv_encoder, but this interface has two shortcomings. First, this interface does not provide padding. Although the effect of padding may be small in one-dimensional convolution due to globalmaxpool, it may be a little different. If I want to pad data, I need to pad the input 3D data independently, but the ndarray pad currently only supports 4d and 5d. I need to change the dimension, which adds some extra unnecessary operations. Maybe adding padding to the implementation of this interface will be easier to use.
Second, I think the input of the general convolution will be the 0th dimension of the batch, but the input of this interface is not. I have seen the implementation. it has internally performed the dimensional transformation. Is this necessary? Maybe this interface will be more intuitive when the batch is in the 0th dimension.